buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java

```diff
@@ -1431,21 +1477,40 @@ public ByteBuf internalComponent(int cIndex) {
      * @param offset the offset for which the {@link ByteBuf} should be returned
      */
     public ByteBuf internalComponentAtOffset(int offset) {
-        return findComponent(offset).buf;
+        return findComponent(offset).slice();
     }
 
+    // weak cache - check it first when looking for component
+    private Component lastAccessed;
+
     private Component findComponent(int offset) {
+        Component la = lastAccessed;
+        if (la != null && offset >= la.offset && offset < la.endOffset) {
+           ensureAccessible();
+           return la;
+        }
         checkIndex(offset);
+        return findIt(offset);
+    }
+
+    private Component findComponent0(int offset) {
+        Component la = lastAccessed;
+        if (la != null && offset >= la.offset && offset < la.endOffset) {
+           return la;
+        }
+        return findIt(offset);
+    }
 
-        for (int low = 0, high = components.size(); low <= high;) {
+    private Component findIt(int offset) {
```

Comment:
this always searches one direction.  can we do better in large componentbuffers by checking which direction would be faster to search in?  also, if lastAccessed is close to the target offset, should we start from there?

Reply:
I'm not sure what you mean by one direction here since it's a binary search. I did experiment a little with choosing the next component based on the target byte location within the remaining range rather than just choosing the middle component each time, but the cost of the additional ops appeared to outweigh any benefits (I expect with more attention _some_ improvement could be made though).

> if lastAccessed is close to the target offset, should we start from there?

@normanmaurer [suggested the same thing](https://github.com/netty/netty/pull/8406#discussion_r226628136). It would complicate the logic since the index would need to be cached separately (which would not be updated atomically with the component ref), or an index field could be added to the `Component` class which seemed overkill just for this purpose. But I'd be happy to experiment with it and see if there's measurable diff in the numbers.

Reply:
I think we can do this as a followup. This PR is already big enough .. @mosesn ok with it ?

Reply:
yep, that's fine.